menpofit/dpm/fitter.py

```diff
@@ -0,0 +1,637 @@
+                    if not latent:
+                        qp.write(ex)
+                        qp.increment_neg_ub(root_score[y, x])
+                        diffs.append(abs(qp.score_neg() - root_score[y, x]))
+
+                if not latent and xs.shape[0] > 0 and qp.n < np.size(qp.a):
+                    # part of the debugging, might need in short future
+                    # print qp.score_neg(), rscore[y, x], qp.score_neg() - rscore[y, x]
+                    # scores1 = qp.slow_score(-(qp.w + qp.w0 * qp.wreg)/qp.cneg, qp.x[:, qp.n-1])
+                    # print 'score1 :', scores1
+                    # scores2 = compute_filters_scores_only(old_scores, tree, filter_ids, ptr)
+                    # print 'score2 :', scores2
+                    assert(qp.score_neg() - root_score[y, x] < 10**-5)
+
+                if not latent and (qp.lb < 0 or 1-qp.lb/qp.ub > 0.05 or qp.n == np.size(qp.sv)):
+                    model = qp.obtimise(model)
+                    filters_all = model.get_filters_weights()
+                    unary_scores_all = convolve_python_f(feat, filters_all)
+                    defs_all = model.get_defs_weights()
+                    anchors_all = model.get_defs_anchors()
+
+        if latent:
+            qp.write(ex)
+            # part of the debugging, might need in short future
+            # scores1 = qp.slow_score((qp.w + qp.w0 * qp.wreg)/qp.cpos, qp.x[:, qp.n-1])
+            # scores2 = compute_filters_scores_only(best_old_scores, best_tree, best_filter_ids, ptr)
+            # assert(qp.score_pos()[qp.n-1] - best_rscore[y, x] < 10**-5)
+            if np.size(boxes) > 0:
+                boxes = boxes[-1]
+        return boxes, model
+
+
+def _test_overlap(fsz, feat, scale, padding, img_size, box, overlap):
+    # return the 2d boolean matrix with the same size of each part score indicating if each position overlap with the
+    # bounding box more than a given threshold
+    [bx1, by1, bx2, by2] = box
+    (_, dim_y, dim_x) = feat.shape
+    (size_x, size_y) = fsz
+    (padx, pady) = padding
+    (imgx, imgy) = img_size
+    x1 = (np.array(range(0, dim_x - size_x + 1)) - padx - 1) * scale + 1
+    y1 = (np.array(range(0, dim_y - size_y + 1)) - pady - 1) * scale + 1
+    x2 = x1 + size_x*scale - 1
+    y2 = y1 + size_y*scale - 1
+    x1 = np.minimum(np.maximum(x1, np.zeros_like(x1)), np.full_like(x1, imgx-1))
+    x2 = np.maximum(np.minimum(x2, np.full_like(x2, imgx-1)), np.zeros_like(x2))
+    y1 = np.minimum(np.maximum(y1, np.zeros_like(y1)), np.full_like(y1, imgy-1))
+    y2 = np.maximum(np.minimum(y2, np.full_like(y2, imgy-1)), np.zeros_like(y2))
+    xx1 = np.maximum(x1, np.full_like(x1, bx1))
+    xx2 = np.minimum(x2, np.full_like(x2, bx2))
+    yy1 = np.maximum(y1, np.full_like(y1, by1))
+    yy2 = np.minimum(y2, np.full_like(y2, by2))
+    w = xx2 - xx1 + 1
+    h = yy2 - yy1 + 1
+    w[w < 0] = 0
+    h[h < 0] = 0
+    inter = np.dot(np.transpose(np.asmatrix(h)), np.asmatrix(w))
```

Comment:
Originally, I had a problem that the variable ```h``` can not be transpose as it is 1D array.
Will fix this by making ```x1, y1``` 2D array